### ISSUE TITLE
ErrorHolder is basically untested

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1946]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#1946|
|Type|defect|
|Created|2006-07-22 14:01:29Z|
|Branch|https://github.com/twisted/twisted/tree/error-holder-tests-1946|

twisted.trial.runner.ErrorHolder has no unit tests.  It should at least have tests to make sure that it duck-types to a TestCase

 * can be run
 * adds an error
 * can be visited

Arises from [#1940](https://github.com/twisted/twisted/issues/1940)

Attachments:

* [errorholder.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b6a/b6a437646b7e39bc70c5549a82d04cfe666dbcd9/c3ce6f544eb38d6dcef9fb3e04038257b16a992a.patch) (2235 bytes) - added by kelly on 2009-11-06 12:53:21Z - Some tests for ErrorHolder.

<details><summary>Searchable metadata</summary>

```
trac-id__1946 1946
type__defect defect
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_error_holder_tests_1946 branches/error-holder-tests-1946
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1153576889000000 1153576889000000
changetime__1317564763000000 1317564763000000
version__None None
owner__exarkun exarkun
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kelly's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|kelly commented|
|-|-|

As a side note t.t.runner.TestHolder says it implements t.t.itrial.ITestCase but doesn't actually fully implement it (missing run, __call__ and countTestCases).  Obviously not causing a problem, but lies hurt.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks!  A couple very minor points, which I'll correct in the patch and then apply:

  1. `assertEquals` is preferred to `assertEqual` in Twisted code.
  1. _Check that_ is somewhat redundant in test method docstrings.  Instead, you can just say what's being checked.

Also, I notice that visit is still uncovered, but since I think visiting in trial is going away, I'm not very worried about that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

(In [27456]) Add unit tests for `twisted.trial.runner.ErrorHolder`

Author: kelly
Reviewer: exarkun
Fixes: [#1946](https://github.com/twisted/twisted/issues/1946)

Add unit tests for `ErrorHolder` which previously had none.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `reopened`|
|-|-|

(In [27457]) Revert r27456 because of Python 2.4 test suite regression

Reopens [#1946](https://github.com/twisted/twisted/issues/1946)

```
===============================================================================
[FAIL]: twisted.trial.test.test_runner.TestErrorHolder.test_repr

Traceback (most recent call last):
  File "/var/lib/buildbot/twisted-bigdogvm1/ubuntu-py2.4-select/Twisted/twisted/trial/test/test_runner.py", line 730, in test_repr
    "<ErrorHolder description='description' "
twisted.trial.unittest.FailTest: not equal:
a = "<ErrorHolder description='description' error=<twisted.python.failure.Failure exceptions.ZeroDivisionError>>"
b = "<ErrorHolder description='description' error=<twisted.python.failure.Failure <type 'exceptions.ZeroDivisionError'>>>"

-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to kelly<br>@exarkun set status to `new`|
|-|-|

If I'd looked at that big ugly string in `test_repr` more carefully in the first place, I probably would have decided that `<type 'exceptions.ZeroDivisionError'>` is the wrong thing to show up there anyway.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [32668]) Branching to 'error-holder-tests-1946'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

Updated this to modern-ish Twisted, made the tests a bit more complete, and changed the implementation to actually behave better (I think).

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/error-holder-tests-1946)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

reviewin'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

This looks great.  It's been a _long_ time since I've seen a build where
_all_ the builds are green - actually, given that this is a change to Trial, I
was highly suspicious that it damaged error reporting somehow ;).

    1. I think that `ErrorHolderTestsMixin.test_countTestCases` means to refer
    to `runner.ErrorHolder.countTestCases`, not `runner.countTestCases`.
    1. `test_repr` refers to `ErrorHolder.__repr__` but everyone else refers to
    `runner.ErrorHolder`; consistency would be good.   For brevity's sake, I
    think it might be better to actually import the name so that it can be
    `ErrorHolder` everywhere.
    1. the 4-line `try: 1 / 0; except: Failure()` stanza is repeated -
    ''almost'' exactly - twice in this code, and I'm sure in a zillion other
    places in testing Trial and Failure.  I am starting to think that this is an
    overly cute way to get a traceback.  It's not that it's unclear, but it
    always makes me take just a second to make sure that there isn't some more
    interesting arithmetic thing happening, and I imagine that would take
    someone a moment longer if they weren't familiar with this idiom already.
    Also, while '`1 / 0`' is shorter than '`raise Something()`', the
    `try:except:something = Failure()` block is longer than `something =
    failWithTraceback()`, especially vertically.  So, at least in the future,
    consider refactoring some of this code to pull the
    artificial-`Failure`-synthesis stuff into a utility function.  (It might be
    interesting to also make such a function create different exception types,
    and maybe vary the length and contents of the traceback, so that it would be
    easier to have test cases that test other than the most trivial possible
    traceback.  That isn't really relevant to this ticket at all though.)
    1. It's great that this formats the exception type correctly now.  That
    prompted me to see what the progress was on getting `getBriefTraceback` to
    do that too.  But I discovered that there is no ticket for that, so I filed
    [#5290](https://github.com/twisted/twisted/issues/5290).  Nothing for you to do here, just link fodder for future
    web-searchers ;-).

Please address to your satisfaction and land.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

in r32732:

  1. Fixed
  2. Fixed (but I left it as `runner.ErrorHolder`)
  3. Yes, it's sort of a sad pattern.  However, the 2nd example needs extra code in the except suite, so it's hard to refactor.  I changed it to not use 1 / 0 to generate the exception, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [32733]) Merge error-holder-tests-1946

Author: kelly, exarkun
Reviewer: review, glyph
Fixes: [#1946](https://github.com/twisted/twisted/issues/1946)

Add some tests for `twisted.trial.runner.ErrorHolder`, which is a test-case like
object for known-in-advance errors and failures.  Also improve the representation
of that object, and normalize its `error` attribute to always be an exc_info-like
object.
